compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/DelegationLowering.kt

```diff
@@ -0,0 +1,237 @@
+import org.jetbrains.kotlin.ir.types.impl.*
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.*
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.types.Variance
+
+internal val propertyDelegationPhase = makeIrFilePhase(
+    ::PropertyDelegationLowering,
+    name = "PropertyDelegation",
+    description = "Construct KProperty instances to be used when accessing delegated properties"
+)
+
+internal class PropertyDelegationLowering(val context: JvmBackendContext) : ClassLoweringPass {
+    private object DECLARATION_ORIGIN_KPROPERTIES_FOR_DELEGATION :
+        IrDeclarationOriginImpl("KPROPERTIES_FOR_DELEGATION")
+
+    private val plainJavaClass =
+        context.getIrClass(FqName("java.lang.Class")).owner
+
+    private val reflectionClass =
+        context.getIrClass(FqName("kotlin.jvm.internal.Reflection")).owner
+
+    private val getOrCreateKotlinClass =
+        reflectionClass.declarations.find { it.descriptor.name == Name.identifier("getOrCreateKotlinClass") }
+
+    private val getOrCreateKotlinPackage =
+        reflectionClass.declarations.find { it.descriptor.name == Name.identifier("getOrCreateKotlinPackage") }
+
+    private fun propertyTypeAndFactory(prefix: String, i: Int) =
+        context.getIrClass(FqName("kotlin.jvm.internal.${prefix.capitalize()}Reference${i}Impl")) to
+                reflectionClass.declarations.find { it.descriptor.name == Name.identifier("${prefix}${i}") }
+
+    private val properties =
+        arrayOf(0, 1, 2).map { propertyTypeAndFactory("property", it) }
+
+    private val mutableProperties =
+        arrayOf(0, 1, 2).map { propertyTypeAndFactory("mutableProperty", it) }
+
+    private val arrayItemGetter =
+        context.ir.symbols.array.owner.functions.single { it.name == Name.identifier("get") }
+
+    private val kPropertyType =
+        context.reflectionTypes.kProperty.toIrType()
+
+    private val kPropertiesFieldType =
+        context.ir.symbols.array.createType(false, listOf(makeTypeProjection(kPropertyType, Variance.OUT_VARIANCE)))
+
+    override fun lower(irClass: IrClass) {
+        // Maintain an array of `KProperty` instances, one per delegated property, as the delegate target is allowed to use identity
+        // comparison of these references for differentiating properties.
+        val kProperties = mutableMapOf<CallableDescriptor, Pair<IrExpression, Int>>()
+        // TODO: annotate as shared-immutable? The old backend does not do so.
+        val kPropertiesField = WrappedFieldDescriptor(Annotations.create(listOf())).let { descriptor ->
+            IrFieldImpl(
+                UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+                DECLARATION_ORIGIN_KPROPERTIES_FOR_DELEGATION,
+                IrFieldSymbolImpl(descriptor),
+                "\$delegatedProperties".synthesizedName,
+                kPropertiesFieldType,
+                Visibilities.PRIVATE,
```

Comment:
In old backend properties are 'final static synthetic'
Syntheticness could be specified via IrDeclarationOriginImpl("KPROPERTIES_FOR_DELEGATION", true)
`TODO` in such case would be redundant or I miss smth?

Reply:
Haven't noticed that IrDeclarationOriginImpl has an additional argument for syntheticness, thanks.

Reply:
Please add TODO that visibility should be package local,
or maybe best switch it temporary to INTERNAL (some inliner test should be passed)  with same TODO